README.md

```diff
@@ -202,6 +202,38 @@ You need a Python 3.7 or later installed on your machine.
 You'll also need the Python PIP and Python VENV modules. On Debian systems, you can install with:
 * `sudo apt install python3-pip python3-venv`
 
+### Optionally Install GDB
+
+You'll only need the GDB utility for on-chip debugging or [for decoding backraces in panics for RISCV Espressif SOCS](#install-esp-idf-monitor).
+
+To install it:
+* Manually download the GDB package matching your OS and Espressif SOC from [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/tools/idf-tools.html#list-of-esp-idf-tools)
+* Unpack (with `tar xvfz`) into a directory of your choice
+* Add the unpacked GDB debugger to your `PATH`, so that e.g. invoking `riscv32-esp-elf-gdb` (or `xtensa-esp-elf-gdb` for xtensa-based SOCs) without any path prefixes is possible, as `esp-idf-monitor` wants that
+
+### Install `esp-idf-monitor`
+
+While `espflash` has a built-in monitor, its monitor is currently unable to properly decode ESP-IDF stacktraces generated by the
```

Comment:
I'm talking about both as the difference between the two is only how they are invoked (one starts monitoring immediately after flashing, while the other just connects to an already flashed instance).

Regardless, in both cases we are talking about _THE_ `espflash` monitor, and this monitor is currently completely incapable of displaying proper stacktraces for _default_ `esp-idf-template` project configuration. I guess baremetal does not care, as they use Rust only, and they have frame-pointers enabled by default for `esp-backtrace` (see below)

There are workarounds like:
- Enabling frame pointers for `rustc`, however this will give you only partial stacktraces for the rust side. Might not even work at all, because ESP IDF panics would still dump stack **memory** and not `Backtrace [PC:SP...]` info as on xtensa. And no, I'm not sure it is even possible to enable FP on GCC (googled, tried, no avail)
- Enabling EH_FRAME support in ESP IDF and paying 20%-100% firmware overhead. This is BTW equivalent to what `esp-idf-monitor` does, except `esp-idf-monitor` is doing it on the host, and thus your firmware stays small.

`esp-idf-monitor` (as we have discussed once) is capable of dealing directly with the chunk of stack memory dumped by default by ESP IDF riscv MCUs in case of ESP IDF panics (crashes), and decoding it properly. 
Essentially, it is taking the stack memory dump, massaging it a bit to look like a coredump, and then handing it off to GDB in headless mode for analysis, where GDB then decodes the stacktrace.


All of this can be implemented in `espflash` I suppose, but in the meantime, using `esp-idf-monitor` is a good workaround.

Reply:
Since there is a small chance of you telling me shortly that `espflash monitor` does decode stacktraces on riscv ( :-) ), and you might refer to its simple algorithm which is grepping `u32` byte sequences which look like program memory, here is:
- First: a properly decoded ESP IDF panic, with `esp-idf-monitor`
- Second: a mindless "grepping" exercise done again by `esp-idf-monitor` when it is NOT instructed to do proper decodes, over the raw stack memory dump generated by the ESP IDF crash; this is equivalent to the only thing `espflash monitor` can do currently.

I find the "decoding" done via "greps" very unreadable, and if you look at what it decoded - plain misleading.

First: `esp-idf-monitor` proper decoding:
```
E(8576) OPENTHREAD:[C] P-SpinelDrive-: Failed to reset co-processor!
E(8576) OPENTHREAD:[C] Platform------: ResetCoprocessor() at spinel_driver.cpp:160: Failure

abort() was called at PC 0x420d80a9 on core 0
--- 0x420d80a9: syscall_not_implemented_aborts at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/newlib/syscalls.c:27

--- Stack dump detected
Core  0 register dump:
MEPC    : 0x40381b4a  RA      : 0x40387f30  SP      : 0x3fcad7d0  GP      : 0x3fc95c00  
--- 0x40381b4a: panic_abort at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/panic.c:463
0x40387f30: __ubsan_include at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/ubsan.c:311

TP      : 0x3fcae3f0  T0      : 0x37363534  T1      : 0x7271706f  T2      : 0x33323130  
S0/FP   : 0x00000004  S1      : 0x3fcad834  A0      : 0x3fcad7fc  A1      : 0x3fcad832  
A2      : 0x00000000  A3      : 0x3fcad829  A4      : 0x00000001  A5      : 0x3fca7000  
A6      : 0x7a797877  A7      : 0x76757473  S2      : 0x3fcad8bc  S3      : 0x00000001  
S4      : 0x3fcad8bc  S5      : 0x3fc9a8ec  S6      : 0x00000000  S7      : 0x00000000  
S8      : 0x00000000  S9      : 0x00000000  S10     : 0x00000000  S11     : 0x00000000  
T3      : 0x6e6d6c6b  T4      : 0x6a696867  T5      : 0x66656463  T6      : 0x62613938  
MSTATUS : 0x00001881  MTVEC   : 0x40380001  MCAUSE  : 0x00000007  MTVAL   : 0x00000000  
--- 0x40380001: _vector_table at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/riscv/vectors_intc.S:54

MHARTID : 0x00000000  


--- Backtrace:


panic_abort (details=details@entry=0x3fcad7fc "abort() was called at PC 0x420d80a9 on core 0") at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/panic.c:463
warning: 463	/home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/panic.c: No such file or directory
#0  panic_abort (details=details@entry=0x3fcad7fc "abort() was called at PC 0x420d80a9 on core 0") at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/panic.c:463
#1  0x40387f30 in esp_system_abort (details=details@entry=0x3fcad7fc "abort() was called at PC 0x420d80a9 on core 0") at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/port/esp_system_chip.c:92
#2  0x403908be in abort () at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/newlib/abort.c:38
#3  0x420d80ac in syscall_not_implemented_aborts () at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/newlib/syscalls.c:27
#4  0x421d4924 in exit (code=1) at /builds/idf/crosstool-NG/.build/riscv32-esp-elf/src/newlib/newlib/libc/stdlib/exit.c:65
#5  0x4214ad1a in ot::Spinel::SpinelDriver::ResetCoprocessor (this=this@entry=0x3fc9a8e8 <s_spinel_driver>, aSoftwareReset=aSoftwareReset@entry=true) at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/lib/spinel/spinel_driver.cpp:160
#6  0x4214b29e in ot::Spinel::SpinelDriver::Init (this=this@entry=0x3fc9a8e8 <s_spinel_driver>, aSpinelInterface=..., aSoftwareReset=aSoftwareReset@entry=true, aIidList=aIidList@entry=0x3fcad8bc "", aIidListLength=aIidListLength@entry=1 '\001') at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/lib/spinel/spinel_driver.cpp:81
#7  0x42118794 in esp_openthread_radio_init (config=<optimized out>) at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/src/port/esp_openthread_radio_spinel.cpp:91
#8  0x42117fa0 in esp_openthread_platform_init (config=0x3fcadd14) at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/src/esp_openthread_platform.cpp:149
#9  0x421165ee in esp_openthread_init (config=<optimized out>) at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/src/esp_openthread.cpp:69
#10 0x420127a0 in esp_idf_svc::thread::ThreadDriver<esp_idf_svc::thread::Host>::init<esp_idf_svc::thread::Host> (self=0x0) at src/thread.rs:1060
#11 0x00000000 in ?? ()
Backtrace stopped: frame did not save the PC
```

Second: "Mindless" grepping with `addr2line` (or whatever xmas-elf equivalent `espmonitor` uses). This can only work on xtensa, which outputs `Backtrace [IP:SP]`, because of the `IP` which is a precise instruction pointer. But I think it produces garbage on raw stack memory:
```
W(8566) OPENTHREAD:[W] P-SpinelDrive-: Wait for response timeout
E(8566) OPENTHREAD:[C] P-SpinelDrive-: Failed to reset co-processor!
E(8566) OPENTHREAD:[C] Platform------: ResetCoprocessor() at spinel_driver.cpp:160: Failure

abort() was called at PC 0x420d80a9 on core 0
--- 0x420d80a9: syscall_not_implemented_aborts at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/newlib/syscalls.c:27

Core  0 register dump:
MEPC    : 0x40381b4a  RA      : 0x40387f30  SP      : 0x3fcad7d0  GP      : 0x3fc95c00  
--- 0x40381b4a: panic_abort at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/panic.c:463
0x40387f30: __ubsan_include at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/esp_system/ubsan.c:311

TP      : 0x3fcae3f0  T0      : 0x37363534  T1      : 0x7271706f  T2      : 0x33323130  
S0/FP   : 0x00000004  S1      : 0x3fcad834  A0      : 0x3fcad7fc  A1      : 0x3fcad832  
A2      : 0x00000000  A3      : 0x3fcad829  A4      : 0x00000001  A5      : 0x3fca7000  
A6      : 0x7a797877  A7      : 0x76757473  S2      : 0x3fcad8bc  S3      : 0x00000001  
S4      : 0x3fcad8bc  S5      : 0x3fc9a8ec  S6      : 0x00000000  S7      : 0x00000000  
S8      : 0x00000000  S9      : 0x00000000  S10     : 0x00000000  S11     : 0x00000000  
T3      : 0x6e6d6c6b  T4      : 0x6a696867  T5      : 0x66656463  T6      : 0x62613938  
MSTATUS : 0x00001881  MTVEC   : 0x40380001  MCAUSE  : 0x00000007  MTVAL   : 0x00000000  
--- 0x40380001: _vector_table at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/riscv/vectors_intc.S:54

MHARTID : 0x00000000  

Stack memory:
3fcad7d0: 0x00000000 0x00000001 0x3fcad830 0x403908be 0x00000000 0x3fc96d54 0x3fca7c18 0x3fc98378
--- 0x403908be: abort at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/newlib/abort.c:35

3fcad7f0: 0x3fcad834 0x3fc98394 0x3fcad830 0x726f6261 0x20292874 0x20736177 0x6c6c6163 0x61206465
3fcad810: 0x43502074 0x34783020 0x38643032 0x20396130 0x63206e6f 0x2065726f 0x00000030 0x421d0000
--- 0x421d0000: icmp6_raw_recv_handler at esp_openthread_infra_if.c.obj:?

3fcad830: 0x00000030 0x64303234 0x39613038 0x00000000 0x00000000 0x00000000 0x00000001 0x420d80ac
--- 0x420d80ac: fcntl at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/newlib/syscalls.c:156

3fcad850: 0x3fcad8bc 0x00000000 0x00000001 0x421d4924 0x3c2327d8 0x00000001 0x3fc9a8e8 0x4214ad1a
--- 0x421d4924: _findenv at /builds/idf/crosstool-NG/.build/riscv32-esp-elf/src/newlib/newlib/libc/stdlib/getenv.c:68
0x4214ad1a: ot::Spinel::SpinelDriver::CheckSpinelVersion() at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/lib/spinel/spinel_driver.cpp:413

3fcad870: 0x7665642f 0x00000001 0x3fc9a8e8 0x4214b29e 0x3c2283ec 0x00000000 0x00000000 0x00000000
--- 0x4214b29e: ot::Spinel::SpinelDriver::Init(ot::Spinel::SpinelInterface&, bool, unsigned char const*, unsigned char) at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/lib/spinel/spinel_driver.cpp:82

3fcad890: 0x00000000 0x3fcadd14 0x3fc9b000 0x42118794 0x42124328 0x42124380 0x421243b8 0x42124358
--- 0x42118794: esp_openthread_radio_init at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/src/port/esp_openthread_radio_spinel.cpp:92
0x42124328: otPlatRadioReceiveDone at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/core/radio/radio_platform.cpp:49
0x42124380: otPlatRadioTxDone at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/core/radio/radio_platform.cpp:86
0x421243b8: otPlatRadioEnergyScanDone at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/core/radio/radio_platform.cpp:109
0x42124358: otPlatRadioTxStarted at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/core/radio/radio_platform.cpp:69

3fcad8b0: 0x00000000 0x421243e0 0x42124404 0x42118300 0x3fcadde8 0x0401dde4 0x00000000 0x42117fa0
--- 0x421243e0: otPlatDiagRadioReceiveDone at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/core/radio/radio_platform.cpp:121
0x42124404: otPlatDiagRadioTransmitDone at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/openthread/src/core/radio/radio_platform.cpp:135
0x42118300: esp_openthread_task_queue_process at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/src/esp_openthread_task_queue.c:99 (discriminator 1)
0x42117fa0: esp_openthread_platform_init at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/src/esp_openthread_platform.cpp:149

3fcad8d0: 0x00000000 0x00000000 0x3c221000 0x421165ee 0x00000000 0x00000000 0x3c221000 0x420127a0
--- 0x421165ee: esp_openthread_init at /home/ivan/dev/esp-idf-svc/.embuild/espressif/esp-idf/v5.3.1/components/openthread/src/esp_openthread.cpp:69
0x420127a0: esp_idf_svc::thread::ThreadDriver<T>::init at /home/ivan/dev/esp-idf-svc/src/thread.rs:1060

3fcad8f0: 0x00000000 0x00000000 0x00000000 0x3c206108 0x3fcadde4 0x3fcadd0c 0x00000000 0x00000000
3fcad910: 0x3fcadde4 0x00000000 0x3fcadd0c 0x00000000 0x3fcadde4 0x04000000 0x3fcadde4 0x01000000
3fcad930: 0x3c206108 0x3fcadde4 0x3c206108 0x3c206108 0x3fcbe4fc 0x3fcbe4fc 0x3fcbe4fc 0x42004aae
--- 0x42004aae: esp_idf_svc::thread::EspThread<T>::init at /home/ivan/dev/esp-idf-svc/src/thread.rs:1476

3fcad950: 0x00000000 0x3fcadcfc 0x00000000 0x3c205600 0x00000000 0x3fcadcfc 0x000003e8 0x42003ec8
--- 0x42003ec8: thread_br::example::main at /home/ivan/dev/esp-idf-svc/examples/thread_br.rs:143

3fcad970: 0xa5a5a5a5 0xa5a5a5a5 0xa5a5a5a5 0xa5a5a5a5 0x3fcae018 0x000000fc 0x3fcadcfc 0x3fcadff0
3fcad990: 0x3fcb2020 0x3fcb2214 0x3fcb2be4 0x3fcadef4 0x3fcadc60 0x3fcadcc4 0x42000052 0x3fcadc1c
--- 0x42000052: <embedded_svc::ipv4::IpInfo as core::fmt::Debug>::fmt at /home/ivan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/embedded-svc-0.28.0/src/ipv4.rs:239

3fcad9b0: 0x3fcadbec 0x3fcadb80 0x00000040 0x3fcada08 0x3fcadb44 0x3fcada88 0x3fcb2020 0x3fcadb04
3fcad9d0: 0x3fcae14c 0xa5a5a5a5 0xa5a5a5a5 0xa5a5a5a5 0xa5a5a5a5 0x00000000 0x3fcb2020 0x00000000
3fcad9f0: 0x3fcb2020 0x3fcb2214 0x00000000 0x3fcb2214 0x3fcb2be4 0x00000000 0x00000000 0x3fcb2020
3fcada10: 0x3fcbb5b4 0x3fcbb5f8 0x3c228140 0x0000000b 0xffffffff 0x3fcbb594 0x3fcb2214 0x3fcbd030
3fcada30: 0x00000000 0x00000000 0x3fcbcd44 0x00000000 0x00000001 0x3fcb2020 0x00000000 0x3fcb2020
3fcada50: 0x3fcbb5b4 0x3fcbb5f8 0x3c228140 0x0000000b 0xffffffff 0x3fcbb594 0x3fcb2214 0x3fcbd030
3fcada70: 0x00000000 0x00000000 0x3fcbcd44 0x00000000 0x00000001 0x3fcb2020 0x00000000 0x3fcb2020
3fcada90: 0x3fcbb5b4 0x3fcbb5f8 0x3c228140 0x0000000b 0xffffffff 0x3fcbb594 0x3fcb2214 0x3fcbd030
3fcadab0: 0x00000000 0x00000000 0x3fcbcd44 0x00000000 0x00000001 0x3fcb2020 0x00000000 0x3fcb2020
3fcadad0: 0x3fcbb5b4 0x3fcbb5f8 0x3c228140 0x0000000b 0xffffffff 0x3fcbb594 0x3fcb2214 0x3fcbd030
3fcadaf0: 0x00000000 0x00000000 0x3fcbcd44 0x00000000 0x00000001 0x00000000 0x3fcb2020 0x3fcbb5b4
3fcadb10: 0x3fcbb5f8 0x3c228140 0x0000000b 0xffffffff 0x3fcbb594 0x3fcb2214 0x3fcbd030 0x00000000
3fcadb30: 0x00000000 0x3fcbcd44 0x00000000 0x00000001 0x3fcb2214 0x00000000 0x3fcb2020 0x3fcbb5b4
3fcadb50: 0x3fcbb5f8 0x3c228140 0x0000000b 0xffffffff 0x3fcbb594 0x3fcb2214 0x3fcbd030 0x00000000
3fcadb70: 0x00000000 0x3fcbcd44 0x00000000 0x00000001 0x00000000 0x3fcb2020 0x3fcbb5b4 0x3fcbb5f8
3fcadb90: 0x3c228140 0x0000000b 0xffffffff 0x3fcbb594 0x3fcb2214 0x3fcbd030 0x00000000 0x00000000
3fcadbb0: 0x3fcbcd44 0x00000000 0x00000001 0x3fcb2020 0x00000000 0x0aa8c001 0x00000101 0xa8c00000
```